staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go

```diff
@@ -0,0 +1,344 @@
+	for {
+		if t.Kind == types.Alias {
+			t = t.Underlying
+		} else if t.Kind == types.Pointer {
+			t = t.Elem
+		} else {
+			break
+		}
+	}
+	return t
+}
+
+func (lttv listTypeTagValidator) Docs() TagDoc {
+	doc := TagDoc{
+		Tag:         lttv.TagName(),
+		Scopes:      lttv.ValidScopes().UnsortedList(),
+		Description: "Declares a list field's semantic type.",
+		Payloads: []TagPayloadDoc{{
+			Description: "<type>",
+			Docs:        "map | atomic",
+		}},
+	}
+	return doc
+}
+
+type listMapKeyTagValidator struct {
+	byFieldPath map[string]*listMap
+}
+
+func (listMapKeyTagValidator) Init(Config) {}
+
+func (listMapKeyTagValidator) TagName() string {
+	return listMapKeyTag
+}
+
+func (listMapKeyTagValidator) ValidScopes() sets.Set[Scope] {
+	return listTagsValidScopes
+}
+
+func (lmktv listMapKeyTagValidator) GetValidations(context Context, _ []string, payload string) (Validations, error) {
+	t := context.Type
+	if t.Kind == types.Alias {
+		t = t.Underlying
+	}
+	if t.Kind != types.Slice && t.Kind != types.Array {
+		return Validations{}, fmt.Errorf("can only be used on list types")
+	}
+	if realType(t.Elem).Kind != types.Struct {
+		return Validations{}, fmt.Errorf("only lists of structs can be list-maps")
+	}
+
+	found := false
+	for _, memb := range realType(t.Elem).Members {
+		//FIXME: should this be the JSON name?  That's a change...
+		if memb.Name == payload {
+			if k := realType(memb.Type).Kind; k != types.Builtin {
+				return Validations{}, fmt.Errorf("only primitive types can be list-map keys, not %s", k)
+			}
+			found = true
+			break
+		}
+	}
+	if !found {
+		return Validations{}, fmt.Errorf("field %q not found in list's value-type", payload)
+	}
+	if lmktv.byFieldPath[context.Path.String()] == nil {
+		lmktv.byFieldPath[context.Path.String()] = &listMap{}
+	}
+	lm := lmktv.byFieldPath[context.Path.String()]
+	lm.keyFields = append(lm.keyFields, payload)
+
+	// This tag doesn't generate any validations.  It just accumulates
+	// information for other tags to use.
+	return Validations{}, nil
+}
+
+func (lmktv listMapKeyTagValidator) Docs() TagDoc {
+	doc := TagDoc{
+		Tag:         lmktv.TagName(),
+		Scopes:      lmktv.ValidScopes().UnsortedList(),
+		Description: "Declares a named sub-field of a list's value-type to be part of the list-map key.",
+		Payloads: []TagPayloadDoc{{
+			Description: "<field-name>",
+			Docs:        "The name of the field.",
+		}},
+	}
+	return doc
+}
+
+type eachValTagValidator struct {
+	byFieldPath map[string]*listMap
+	validator   Validator
+}
+
+func (evtv *eachValTagValidator) Init(cfg Config) {
+	evtv.validator = cfg.Validator
+}
+
+func (eachValTagValidator) TagName() string {
+	return eachValTag
+}
+
+var listTagsValidScopes = sets.New(ScopeAny)
+
+func (eachValTagValidator) ValidScopes() sets.Set[Scope] {
+	return listTagsValidScopes
+}
+
+// LateTagValidator indicatesa that validator has to run after the listType and
+// listMapKey tags.
+func (eachValTagValidator) LateTagValidator() {}
```

Comment:
Long term, I like your idea of having logical groups of tags (i.e. structurally informing types like listType being a seaprate group).  This seems fine for now until we build that out.

Reply:
Yeah, this is kind of dirty but it was easy and didn't impact all the other tags.